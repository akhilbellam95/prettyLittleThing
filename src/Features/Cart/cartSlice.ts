import { createSlice } from '@reduxjs/toolkit';

interface CartItem {
  id: number;
  quantity: number;
};

export interface CartState {
  cartItems: CartItem[],
  total: number
}

export const cartInitialState: CartState = {
  cartItems: [] as CartItem[],
  total: 0,
};

export const cartSlice = createSlice({
  name: 'cart',
  initialState: cartInitialState,
  reducers: {
    addToCart: (state, action) => {
      const { id } = action.payload;
      const existingItem = state.cartItems.find((item) => item.id === id);
      let total = state.total;
      if (existingItem) {
        existingItem.quantity++;
      } else {
        state.cartItems.push({ id, quantity: 1 });
      }
      total += action.payload.price;
      state.total = parseFloat(total.toFixed(2));
    },
    removeFromCart: (state, action) => {
      const itemIndex = state.cartItems.findIndex((el) => el.id === action.payload.id);
      if (itemIndex !== -1) {
        const item = state.cartItems[itemIndex];
        let total = state.total;
        if (item.quantity === 1) {
          state.cartItems.splice(itemIndex, 1);
        } else {
          item.quantity--;
        }
        total -= action.payload.price;
        state.total = parseFloat(total.toFixed(2));
      }
    },
    updateTotal: (state, action) => {
      state.total = action.payload;
    },
  },
});

export const { addToCart, removeFromCart, updateTotal } = cartSlice.actions;

export default cartSlice.reducer;
