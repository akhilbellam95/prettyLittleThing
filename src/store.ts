// eslint-disable-next-line prettier/prettier
import { configureStore } from '@reduxjs/toolkit';
import cartReducer from './Features/Cart/cartSlice';
import productReducer from './Features/Product/productSlice';

export const store = configureStore({
  reducer: {
    cart: cartReducer,
    product: productReducer,
  },
  devTools: process.env.NODE_ENV !== 'production',
});

export type AppDispatch = typeof store.dispatch;

export type RootState = ReturnType<typeof store.getState>;
