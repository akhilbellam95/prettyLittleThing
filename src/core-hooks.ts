import { configureStore, createStore } from "@reduxjs/toolkit";
import { useSelector } from "react-redux";
import { TypedUseSelectorHook, useDispatch } from "react-redux";
import { CartState } from "./Features/Cart/cartSlice";
import { ProductsState } from "./Features/Product/productSlice";

export interface ApplicationState {
  cart: CartState;
  product: ProductsState;
}

export const useAppDispatch = () =>
  useDispatch<ReturnType<typeof configureStore>['dispatch']>();

export const useAppSelector: TypedUseSelectorHook<ApplicationState> =
  useSelector;