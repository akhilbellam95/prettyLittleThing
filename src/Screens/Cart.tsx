import { StyleSheet, Text, View } from 'react-native';
import React from 'react';
import CartItems from '../Components/CartItems';
import { ApplicationState, useAppSelector } from '../core-hooks';

const Cart = () => {
  const { cartItems } = useAppSelector((state: ApplicationState) => state.cart);

  return (
    <View style={styles.container}>
      {cartItems.length > 0 ? <CartItems /> : <Text>Cart is empty</Text>}
    </View>
  );
};

export default Cart;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
});
