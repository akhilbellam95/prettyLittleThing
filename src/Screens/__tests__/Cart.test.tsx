import { render } from '@testing-library/react-native';
import React from 'react';
import { Text } from 'react-native';
import { Provider } from 'react-redux';
import { store } from '../../store';
import Cart from '../Cart';
import { useAppSelector } from '../../core-hooks';

jest.mock('../../core-hooks', () => ({
  useAppSelector: jest.fn(),
  useAppDispatch: jest.fn(),
}));

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: jest.fn(),
}));

describe('Home component testing', () => {
  const mockSelector = jest.fn();

  beforeAll(() => {
    (useAppSelector as jest.Mock).mockImplementation(mockSelector);
  });

  it('renders correctly', () => {
    mockSelector.mockReturnValue({
      cartItems: [{ id: 1, quantity: 2 }],
    });

    const snap: unknown = render(
      <Provider store={store}>
        <Cart />
      </Provider>,
    ).toJSON();
    expect(snap).toMatchSnapshot();
  });

  it('renders correctly when cart is empty', () => {
    mockSelector.mockReturnValue({
      cartItems: [],
    });

    const snap: unknown = render(
      <Provider store={store}>
        <Text>Cart is empty</Text>
      </Provider>,
    ).toJSON();
    expect(snap).toMatchSnapshot();
  });
});
