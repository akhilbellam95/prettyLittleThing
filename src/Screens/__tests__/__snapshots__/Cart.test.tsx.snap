// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component testing renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      {
        "flex": 1,
        "width": "90%",
      }
    }
  >
    <View
      style={
        {
          "flex": 6,
        }
      }
    >
      <View
        style={
          {
            "height": 30,
          }
        }
      />
      <Text
        style={
          {
            "fontWeight": "bold",
            "marginBottom": 20,
          }
        }
      >
        Cart Items (1)
      </Text>
      <RCTScrollView
        data={
          [
            {
              "id": 1,
              "quantity": 2,
            },
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={[]}
        viewabilityConfigCallbackPairs={[]}
      >
        <View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View />
          </View>
        </View>
      </RCTScrollView>
    </View>
    <View
      style={
        {
          "flex": 1,
          "flexDirection": "row",
          "marginTop": 20,
        }
      }
    >
      <Text
        style={
          {
            "textAlign": "center",
            "width": 150,
          }
        }
      >
        Cart Total
      </Text>
      <Text />
    </View>
  </View>
</View>
`;

exports[`Home component testing renders correctly when cart is empty 1`] = `
<Text>
  Cart is empty
</Text>
`;
